tests/trainer/test_trainer.py

```diff
@@ -19,15 +19,69 @@
 from tests.base import EvalModelTemplate
 
 
+def test_auto_hparams(tmpdir):
+    class SubClassEvalModelTemplate(EvalModelTemplate):
+        def __init__(self, subclass_arg=1200):
+            super().__init__()
+
+    class SubSubClassEvalModelTemplate(SubClassEvalModelTemplate):
+        pass
+
+    classes = [SubClassEvalModelTemplate, EvalModelTemplate, SubSubClassEvalModelTemplate]
+
+    for CLASS in classes:
+        # test that the model automatically sets the args passed into init as attrs
+        model = CLASS()
+        assert model.batch_size == 32
+        model = CLASS(batch_size=179)
+        assert model.batch_size == 179
+
+        if isinstance(model, SubClassEvalModelTemplate):
+            assert model.subclass_arg == 1200
+
+        # verify that the checkpoint saved the correct values
+        trainer = Trainer(max_steps=20)
+        trainer.fit(model)
+        raw_checkpoint_path = os.listdir(trainer.checkpoint_callback.dirpath)
+        raw_checkpoint_path = [x for x in raw_checkpoint_path if '.ckpt' in x][0]
+        raw_checkpoint_path = os.path.join(trainer.checkpoint_callback.dirpath, raw_checkpoint_path)
+        raw_checkpoint = torch.load(raw_checkpoint_path)
+        assert 'module_arguments' in raw_checkpoint
+        assert raw_checkpoint['module_arguments']['batch_size'] == 179
+
+        # verify that model loads correctly
+        model = CLASS.load_from_checkpoint(raw_checkpoint_path)
+        assert model.batch_size == 179
+
+        # verify that we can overwrite whatever we want
+        model = CLASS.load_from_checkpoint(raw_checkpoint_path, batch_size=99)
+        assert model.batch_size == 99
```

Comment:
@justusschock added this test. Any cases missing?

Reply:
Yes, you only pass yaml serializable stuff in there. For example there may be users who pass their loss functions that way if they experiment with then, but you can't serialize stuff like torch.nn.MSELoss with yaml